README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# PooledPrevalence

<!-- badges: start -->

<!-- badges: end -->

A set of tools to help design a laboratory based prevalence study
harnessing the pooling of laboratory samples to increase resource
efficiency. The tools provide guidance in choosing the number of samples
to pool and the total number of tests to perform, focusing on achieving
the maximum estimation performance (low uncertainty and estimation
error) while keeping the total amount of tests low. In the package there
are functions to simulate study outcomes, optimize the study design, and
finally, analyze the results of a pooled study to retrieve the
prevalence estimates. The package is based on a Bayesian framework, with
the possibility to perform a full hierarchical analysis including
uncertainty due to the sensitivity in acquiring the testing material.
The package was developed at the occasion of the COVID-19 pandemic, but
can be utilized in every laboratory based prevalence study.

This package is produced as a companion to the following guidance from
the European Center for Disease Prevention and Control:
<https://www.ecdc.europa.eu/sites/default/files/documents/Methodology-estimating-point-prevalence%20-SARS-CoV-2-infection-pooled-RT-PCR-testing.pdf>.

## Installation

You can install the released version of PooledPrevalence with:

``` r
# install.packages("devtools")
devtools::install_github("EU-ECDC/PooledPrevalence")
```

## Basic usage

The main goal of the package is to estimate prevalence in a
population/risk group, given the results of a pooled laboratory test.
Let’s assume a test with the following design: 2000 individual
biological samples divided in 10 samples per testing pool (pool size),
for a total of 200 pools. In this experiment we observe 30 positive
pools (15%). With `get_estimates()` it is possible to estimate the
underlying prevalence:

``` r
library(PooledPrevalence)

results <- get_estimates(10, 200, 30)

print(results$estimates)
#>   p_test  k        Est         Lo         Up
#> 1   0.15 30 0.01610737 0.01103635 0.02251263
```

The object `results` contains both the estimates and, if a Bayesian
method is used, the posterior samples of the distribution. The samples
allows the inspection of the full posterior distribution of the
prevalence.

``` r

results <- get_estimates(10, 200, 30)

plot(density(results$samples))
```

<img src="man/figures/README-plot_posterior-1.png" width="100%" />

With Hierarchical Bayesian (HB) estimation is possible to account for
the risk of false negatives due to incorrect sampling of material.
Notice that the resulting estimates are therefore slightly higher. HB
allows also a more correct evaluation of the uncertainty and should be
the preferred method for the final analyses.

``` r

results <- get_estimates(10, 200, 30, method = 'HB')
#> 
#> 
#> @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
#> The Metropolis acceptance rate was 0.75312
#> @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

plot(density(results$samples))
```

<img src="man/figures/README-Hierarchical_Bayes-1.png" width="100%" />

## Hypothesis testing

Having access to the posterior samples allows Bayesian hypothesis
testing for decision making. For example, a public health authority may
want to remove in a geographical region a certain control measure only
if the average prevalence is below 2% with a \> 90% probability:

``` r

results <- get_estimates(10, 200, 30, iters = 200000)

mean(results$samples < .02)
#> [1] 0.892735
```

In this case the probability of a mean prevalence below 2% is 89.3%,
therefore there’s not enough evidence in favor of removing the control
measure. We may want to increase the sample size of the study or wait
more. The Bayesian framework allows to simply add new samples to the
already collected ones and simply update the analysis.

``` r

old.results <- get_estimates(10, 200, 30, iters = 200000)

new.tested.pools <- 60

new.positve.pools <- 9

new.results <- get_estimates(10, 200 + new.tested.pools, 30 + new.positve.pools, iters = 200000)

mean(new.results$samples < .02)
#> [1] 0.92148
```

The new probability is 92.1%, so we have enough evidence to remove the
control measure.

Finally, we want to ascertain the probability that region A, with 30
positive pools out of 200, has a lower prevalence or region B (38
positive pools out of 180).

``` r

region.A <- get_estimates(10, 200, 30, iters = 200000)

region.B <- get_estimates(10, 180, 38, iters = 200000)

mean(region.A$samples < region.B$samples)
#> [1] 0.94025
```

## Study design optimization

The package provides a tool to help program the most efficient design
for a prevalence study based on pooled testing. The methodology tries to
maximize statistical accuracy while increasing the ratio between pool
size and number of tests.

The user needs to specify the operational limits that are driven by the
available resources (number of sampled individuals `n.max` and maximum
number of laboratory test to perform `w.max`), the expected (maximal)
prevalence, and by the possible reduction in sensitivity due to an
excessive number of samples in the same pool (`s.max`). This number
depends on a number of factors (the pathogen, the sensitivity of the
test, the laboratory infrastructure and technique), therefore it needs
to be decided after appropriate local validation tests. A good starting
point for validation is between 10 and 20 samples per pool.

In the following example we investigate the possible designs for a study
with a maximal pool size of 15 and a maximal number of tests and
individuals to sample equal to 2000. We also set a maximal expected
prevalence of 5%. Given the nature of pooled testing, the statistical
accuracy decreases with the prevalence, so it is better to select a
value which is higher than the prevalence we actually expect to find, in
order not to end up with an underpowered study.

``` r

grid <- design_optimization(s.max = 15, w.max = 2000, p = .05, n.max = 2000)

plot_optimization_grid(grid)
```

<img src="man/figures/README-optimization_grid-1.png" width="100%" />

The plot shows the “optimization windows” with different colors, ordered
by the average score (lower is better).

We can observe that the best window (score: -10.89) is achieved with a
pool size `s` of 10 or more and with a number of test `w` higher than
108.

Once a window has been identified, simulation tests can be used to
evaluate the statistical performance. The tests need to be carried on
specific couples of values of pool size `s` and number of tests `w`.

In order to decrease the risk of sensitivity problems we suggest to
evaluate the more conservative designs in term of pool size (lower `s`),
but validation studies can allow to choose a higher value. Regarding the
number of tests, one should use the higher number that is achievable
during the study, since the higher the number of tests, the higher the
statistical precision; through pooling the maximum number of test needed
is already decreased by a factor equal to `s` (e.g. for `s == 10`, given
2000 individuals planned for sampling, at maximum `2000 / s = 200` tests
will be done).

Let’s compare the results in term of statistical accuracy using `w
= 109` and `w = 200`.

``` r

simulation.low_w <- simulate_pool_test(s = 10, w = 108, p = 0.05)

print(format_simulation(simulation.low_w))
#>                            [,1]                           
#> Estimated prevalence       "4.9%, 95%CI: [3.6%, 6.5%]"    
#> Estimation uncertainty     "2.9%, 95%CI: [2.4%, 3.5%]"    
#> Estimation error           "0.48%, 95%CI: [5e-04, 1.7%]"  
#> Unpooled study prevalence  "5%, 95%CI: [3.7%, 6.3%]"      
#> Unpooled study uncertainty "2.6%, 95%CI: [2.3%, 2.9%]"    
#> Unpooled study error       "0.46%, 95%CI: [2.7e-05, 1.5%]"

simulation.high_w <- simulate_pool_test(s = 10, w = 200, p = 0.05)

print(format_simulation(simulation.high_w))
#>                            [,1]                           
#> Estimated prevalence       "5%, 95%CI: [4%, 6.2%]"        
#> Estimation uncertainty     "2.2%, 95%CI: [1.9%, 2.5%]"    
#> Estimation error           "0.38%, 95%CI: [0.021%, 1.2%]" 
#> Unpooled study prevalence  "5%, 95%CI: [4.1%, 6%]"        
#> Unpooled study uncertainty "1.9%, 95%CI: [1.7%, 2.1%]"    
#> Unpooled study error       "0.35%, 95%CI: [1.5e-05, 1.1%]"
```

It is possible to see that in both simulations the statistical accuracy
of the pooled design is similar to that achieved through an unpooled
study (e.g. `s_unpooled = 1` and `w_unpooled = w * s`). Instead, it’s
possible to observe a definite decrease in uncertainty and error passing
from 108 to 200 tests. In any case, both numbers are definitely lower
than than the 2000 tests that would have been performed wihout pooling.


NEWS.md
# PooledTest 0.1

* First commit!

# PooledTest 1.0

* Added optimization algorithm.
* Added hierarchical Bayesian estimation.
* Added proper Readme.
* Numberous bug fixing and implementation of helpers.

# PooledTest 1.0.1

* Fixed a bug into evaluate_beta_params(). cut.plot.at argument was not implemented.

# PooledTest 1.0.2

* Added ... to get_beta_params() to pass arguments to evaluate_beta_params()
* Fixed a bug into evaluate_beta_params() that was preventing the visualization or vertical lines on the distribution boundaries
* Fixed a bug into simulate_pool_test(), the estimation method was not really passed to get_estimates() and the previous method labeling was used.
* In extreme cases (e.g. prevalence = 50%) design_optimization() may give -Inf scores which crash the decision tree algorithm. Set those values to 10 times the minum value. Triggered also a warning.
* If no optimization windows are found by design_optimization() a meaningfull error is raised. 


R/Helpers.R
#' Estimate Beta distibution parameters given a mean and uncertainty boundaries
#'
#' This tool may be helpful in building priors based on the Beta distribution.
#' The user need to pass the expected mean/median of the distibution and the
#' uncertainty boundaries. Then an optimizion algorithm is used to the find the
#' alpha and beta parameters of the beta that reproduce a shape as similar as
#' possible to the expected one. The method doesn't assure perfect coverage,
#' either for optimization limits or because the expected shape is not allowed
#' by the Beta distribution, but is a starting point for further exploration of
#' useful alfa and beta parameters.
#'
#' @param min.val Expected lower uncertainty bound of the ditribution.
#' @param mid.val Expected central value. Can represent the mean, the median or
#'   the mode of the distribution according to the \code{mid.val.type}
#'   parameter.
#' @param max.val Expected upper uncertainty bound of the ditribution.
#' @param levels Uncertainty level corresponding to the \code{min.val} and
#'   \code{max.val} parameters.
#' @param mid.val.type Define which parameter of the distribution \code{mid.val}
#'   refers to, either the mean, the median or the mode.
#' @param force.proper Discourage values of alpha and beta < 1 during the
#'   optimization process. Tests should be done turning this option on and off
#'   and evaluating the results
#' @param ... Pass parameters to the \code{\link{evaluate_beta_params}}
#'   function, like \code{plot = TRUE} or \code{cut.plot.at} to modify y axis
#'   limits.
#'
#' @seealso \code{\link{evaluate_beta_params}}
#'
#' @return A list with the proposed alpha and beta parameters, a description of
#'   the resulting distribution through \code{\link{evaluate_beta_params}}, and
#'   the actual output of the optimization algorithm.
#'
#' @export
#' @importFrom stats qbeta optim
#'
#' @examples
#'
#' # Estimate the alpha and beta parameter of a Beta distribution with 95%
#' # boundaries near 0.0001 and 0.1 and mean value 0.015
#'
#' get_beta_params(.0001, .015, .1, force.proper = TRUE, plot = TRUE)
#'
#' # Same inputs but now force.proper is FALSE
#'
#' get_beta_params(.0001, .015, .1, force.proper = FALSE, plot = TRUE)
#'

get_beta_params <- function(min.val, mid.val, max.val, levels = c(.025, .975), mid.val.type = c('mean', 'median', 'mode'), force.proper = TRUE, ...) {
	obj.fun <- function(params, q = levels, intended = c(min.val, mid.val, max.val)) {
		a = exp(params[1])
		b = exp(params[2])

		med.pred <- if (mid.val.type[1] == 'mean') {
			a / (a + b)
		} else if (mid.val.type[1] == 'median') {
			qbeta(.5, a, b)
		} else if (mid.val.type[1] == 'mode') {
			(a - 1) / (a + b + 2)
		}

		predicted <- c(qbeta(q, a, b), med.pred) %>% log

		res <- sum((log(intended) - predicted)^2)

		if (is.finite(res) & !is.nan(res) & if (force.proper) (a >= 1 & b >= 1) else TRUE) res else Inf
	}

	res <- optim(par = c(0, 0), fn = obj.fun, method = 'Nelder-Mead')

	list(
		alpha = res$par[1] %>% exp,
		beta = res$par[2] %>% exp,
		distr.parameters = evaluate_beta_params(res$par[1] %>% exp, res$par[2] %>% exp, ...),
		opt.results = res
	)
}


#' Inspect the shape of a beta distribution described by the \eqn{\alpha} and
#' \eqn{\beta} arguments
#'
#' This function is useful in defining priors based on the beta distribution.
#' Given the \eqn{\alpha} and \eqn{\beta} parameters, it returns relevant
#' statistics of the distribution and boundary points to help understand its
#' characteristics. Optionally the full PDF of the distribution is plotted.
#'
#' @param alpha \eqn{\alpha} Shape parameter of the Beta distribution.
#' @param beta \eqn{\beta} Shape parameter of the Beta distribution.
#' @param lower.bound Lower bound of the distribution to report.
#' @param upper.bound Upper bound of the distribution to report.
#' @param plot Whether to plot the PDF of the distribution.
#' @param cut.plot.at Since distribution defined by \code{alpha} or \code{beta}
#'   < 1 are strongly concentrated at 0 or 1, squeezing down the rest of the PDF
#'   in the plot, it may be useful to cut the y axis in order to show the
#'   interesting parts of the distribution.
#'
#'
#' @return A vector with the lower/upper bound chosen by the user, the median,
#'   the mean, the standard deviation and the mode of the distribution. If
#'   \code{alpha} or \code{beta} are below 1, the mode cannot be calculated
#'   through \code{(alpha - 1)/(alpha + beta - 2)} so it is approximated through
#'   grid search
#'
#' @export
#'
#' @importFrom magrittr set_names
#'
#' @examples
#'
#' # Alpha and beta chosen in order to have the mean
#' # alpha/(alpha + beta) equal to .05 and precision alpha + beta
#' # equal to 30
#'
#' evaluate_beta_params(.05 * 30, (1 - .05) * 30)
#'
#' # This time the alpha and beta are chosen to have a mode
#' # (alpha - 1)/(alpha + beta - 2) equal to 0.5, with the same precision
#'
#' evaluate_beta_params(.05 * (30 - 2) + 1, (1 - .05) * (30 - 2) + 1)
#'
#' # If we again center the mean at 0.05 but this time the precision is lowered to 15,
#' # alpha goes below 1 and this produces a distribution with most of the mass
#' # at zero. The mode value calculated analytically in this case would be wrong (negative)
#' # so the empirical mode is shown as an approximation
#'
#' evaluate_beta_params(.05 * 15, (1 - .05) * 15)
#'
#'
#'

evaluate_beta_params <- function(alpha, beta, lower.bound = .025, upper.bound = .975, plot = T, cut.plot.at = NA) {

	mid <- .5

	a <- alpha
	b <- beta

	precis <- .000001

	minq <- qbeta(precis, a, b)
	maxq <- qbeta(1 - precis, a, b)

	grid <- dplyr::tibble(
		prev = seq(minq, maxq, 0.0001),
		like = dbeta(prev, a, b),
		post = like/sum(like)
	)

	params <- c(
		qbeta(c(lower.bound, mid, upper.bound), a, b) %>% set_names(c(lower.bound, mid, upper.bound) %>% percent),
		mean = a/(a+b),
		sd = a * b / ((a + b)^2 * (a + b + 1)),
		mode = if (a < 1 | b < 1) grid$prev[which.max(grid$like)] else (a - 1)/(a + b - 2)
	)

	if (plot) {
		p <- ggplot(grid, aes(prev, post)) +
			geom_line() +
			geom_vline(xintercept = c(params[c(1, 3)]), linetype = 'dashed', alpha = .8) +
			geom_point(
				data = data.frame(
					val = params[-5],
					like = dbeta(params[-5], a, b),
					col = c('Interval', 'Median', 'Interval', 'Mean', 'Mode')
				),
				aes(x = val, y = like/sum(grid$like), color = col), size = 3, alpha = .5) +
			# geom_point(aes(x = params['50%'], y = dbeta(params['50%'], a, b), color = 'Median'), size = 3, alpha = .5) +
			# geom_point(aes(x = params['mode'], y = dbeta(params['mode'], a, b), color = 'Mode'), size = 3, alpha = .5) +
			# geom_point(aes(x = params['2.5%'], y = dbeta(params['2.5%'], a, b), color = 'Interval'), size = 3, alpha = .5) +
			# geom_point(aes(x = params['97.5%'], y = dbeta(params['97.5%'], a, b), color = 'Interval'), size = 3, alpha = .5) +
			scale_x_continuous(labels = percent) +
			scale_y_continuous(labels = percent) +
			coord_cartesian(xlim = c(NA, maxq), ylim = c(NA, cut.plot.at)) +
			theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
			labs(x = 'Value', y = 'PDF', color = NULL) +
			options("PooledPrevalence.ggtheme")

		print(p)
	}

	params
}
